### PR TITLE
[SPARK-32192][SQL] Print column name when throws ClassCastException

### PR DESCRIPTION
**Pull request by StefanXiepj**
### What changes were proposed in this pull request?

When somebody changed the type of partition's field, spark will throw ClassCastException. For example, we have a table like this:
```
drop table if exists cast_exception_test;

create table cast_exception_test(c1 int, c2 string) partitioned by (dt string) stored as orc;

insert into table cast_exception_test partition(dt='2020-04-08') values('1', 'jeff_1');
```

If you change the field's type in hive, query the old partition, spark will throw ClassCastException, but hive will not:
```
-- change the field's type using hive
alter table cast_exception_test change column c1 c1 string;
-- hive correct,  but spark throws ClassCastException
select * from cast_exception_test where dt='2020-04-08';
```

### Why are the changes needed?

When the table has many fields, we don's known which field has been changed. If we print out log about this exception, it will very helpful for us to troubleshoot.

### Does this PR introduce _any_ user-facing change?

When the ClassCastException is caused by changed field's type, you can search which field has problem in exexutor logs:
```
20/04/09 17:22:05 ERROR hive.HadoopTableReader: Exception thrown in field <c1>
```

### How was this patch tested?

First, prepare the test data, the table is partitioned and stored as orc:
```
drop table if exists cast_exception_test;
create table cast_exception_test(c1 int, c2 string) partitioned by (dt string) stored as orc;
insert into table cast_exception_test partition(dt='2020-04-08') values('1', 'jeff_1');
``` 

Then, change the field's type in hive. 
```
alter table cast_exception_test change column c1 c1 string;
```

Now the metadata of the table has been modified, but the partition's metadata which is stored in orc file or hive metastore's mysql is still old. So, query command throws ClassCastException in spark, because spark use table's metadata which is different from orc file's metadata. But hive use partition's metadata which is the same as orc file's metadata.

If you query the old partition, spark will thrown ClassCastException, but hive will not:
```
select * from cast_exception_test where dt='2020-04-08';
```

### CODE REVIEW

**Comment by maropu** [comment]
Thanks for your contribution, @StefanXiepj ! btw, which Spark version you used? I think the current master does not accept the alter command;
```
scala> sql("""alter table cast_exception_test change column c1 c1 string""")
org.apache.spark.sql.AnalysisException: ALTER TABLE CHANGE COLUMN is not supported for changing column 'c1' with type 'IntegerType' to 'c1' with type 'StringType';
```

**Comment by StefanXiepj** [comment]
> Thanks for your contribution, @StefanXiepj ! btw, which Spark version you used? I think the current master does not accept the alter command;
> 
> ```
> scala> sql("""alter table cast_exception_test change column c1 c1 string""")
> org.apache.spark.sql.AnalysisException: ALTER TABLE CHANGE COLUMN is not supported for changing column 'c1' with type 'IntegerType' to 'c1' with type 'StringType';
> ```

Sorry, I forgot it, alter command executed in hive, and query command executed in spark.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 1]
```diff
@@ -486,13 +488,20 @@ private[hive] object HadoopTableReader extends HiveInspectors with Logging {
       var i = 0
       val length = fieldRefs.length
       while (i < length) {
-        val fieldValue = soi.getStructFieldData(raw, fieldRefs(i))
-        if (fieldValue == null) {
-          mutableRow.setNullAt(fieldOrdinals(i))
-        } else {
-          unwrappers(i)(fieldValue, mutableRow, fieldOrdinals(i))
+        Try {
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:491]: I don't feel strongly about it, but I prefer simple try-catch myself. I think it avoids having to handle the 'return value' here, when there isn't one within the while loop?

Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:491]: +1

**Comment by maropu** [comment]
> Sorry, I forgot it, alter command executed in hive, and query command executed in spark.

Could you update the PR description and add some tests for the issue?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 2]
```diff
@@ -47,6 +48,7 @@ import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.unsafe.types.UTF8String
 import org.apache.spark.util.{SerializableConfiguration, Utils}
 
+
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:51]: nit: please remove the unnecessary change.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 1 above)
Comment (StefanXiepj) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: get it & done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 2 above)
Comment (StefanXiepj) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: done

**Comment by StefanXiepj** [comment]
> > Sorry, I forgot it, alter command executed in hive, and query command executed in spark.
> 
> Could you update the PR description and add some tests for the issue?

done

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 3]
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.sql.hive
 import java.util.Properties
 
 import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:23]: Shall we remove this, @StefanXiepj ?

Comment (StefanXiepj) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Thanks for your review, i have removed it.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
Merged to master
